### PR TITLE
LED Output Thread

### PR DESCRIPTION
**Pull request by andyp1per**
This puts LED output on a separate thread allowing much longer LED length without dshot interference

### CODE REVIEW

**Comment by andyp1per** [comment]
Tested on a Flywoo AIO - LEDs work fine

File: libraries/AP_HAL_ChibiOS/Scheduler.cpp [hunk 1]
```diff
@@ -105,6 +111,15 @@ void Scheduler::init()
                      this);                     /* Thread parameter.    */
 #endif
 
+#ifndef HAL_NO_LED_THREAD
+    // setup the LED thread - this will call tasks at 1kHz
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/Scheduler.cpp:115]: this is starting the thread even on boards with no LEDs of this type. Most boards don't have them, so why not create dynamically to save memory/cpu?

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL_ChibiOS/Scheduler.cpp (hunk 1 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/Scheduler.cpp]: Fixed

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp [hunk 2]
```diff
@@ -145,6 +149,51 @@ void RCOutput::init()
...
+#ifdef HAL_NO_LED_THREAD
+    return false;
+#else
+    WITH_SEMAPHORE(led_thread_sem);
+
+    if (led_thread_ctx) {
+        return true;
+    }
+
+    if (!hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&RCOutput::led_thread, void), "led", LED_THD_WA_SIZE, AP_HAL::Scheduler::PRIORITY_LED, 0)) {
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/RCOutput.cpp:164]: if we have two groups with LEDs we could end up with two threads, as there is a race in the settings of led_thread_ctx

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL/RCOutput.h [hunk 3]
```diff
@@ -222,6 +222,9 @@ class AP_HAL::RCOutput {
     // static to allow use in the ChibiOS thread stuff
     static bool is_dshot_protocol(const enum output_mode mode);
 
+    static bool is_led_protocol(const enum output_mode mode) {
+      return mode == MODE_NEOPIXEL || mode == MODE_PROFILED;
```
Comment (peterbarker) [inline libraries/AP_HAL/RCOutput.h:226]: This should be a switch statement so if we get a new protocol the user has to add it in here.

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp [hunk 4]
```diff
@@ -49,6 +49,10 @@ extern AP_IOMCU iomcu;
 #endif
 
 #define RCOU_SERIAL_TIMING_DEBUG 0
+#define LED_THD_WA_SIZE 256
+#if !defined(HAL_NO_LED_THREAD) && defined(HAL_NO_RCOUT_THREAD)
+#define HAL_NO_LED_THREAD
```
Comment (peterbarker) [inline libraries/AP_HAL_ChibiOS/RCOutput.cpp:54]:
```suggestion
#define HAL_NO_LED_THREAD 1
```

... also, make sure this is always defined and stop checking for definition elsewhere, check for boolean value.  This prevents bugs where we have different include orders leading to different values for this define

**Review by peterbarker — DISMISSED** [review]

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp [hunk 5]
```diff
@@ -1250,6 +1303,27 @@ void RCOutput::timer_tick(uint64_t time_out_us)
     }
 }
 
+/*
+  periodic timer called from led thread. This is used for LED output
+ */
+void RCOutput::led_timer_tick(uint32_t time_out_us)
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/RCOutput.cpp:1309]: time_out_us needs to be uint64_t

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp [hunk 6]
```diff
@@ -1250,6 +1303,27 @@ void RCOutput::timer_tick(uint64_t time_out_us)
...
+{
+    if (serial_group) {
+        return;
+    }
+
+    // if we have enough time left send out LED data
+    if (serial_led_pending && (time_out_us > (AP_HAL::micros() + (LED_OUTPUT_PERIOD_US >> 1)))) {
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/RCOutput.cpp:1316]: looks like 32 bit wrap bug

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp [hunk 7]
```diff
@@ -2426,6 +2500,13 @@ void RCOutput::serial_led_send(const uint16_t chan)
         return;
     }
 
+    {
+        WITH_SEMAPHORE(led_thread_sem);
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/RCOutput.cpp:2504]: don't need the semaphore

File: libraries/AP_HAL_ChibiOS/RCOutput.h [hunk 8]
```diff
@@ -104,6 +104,11 @@ class ChibiOS::RCOutput : public AP_HAL::RCOutput
      */
     void timer_tick(uint64_t last_run_us);
 
+    /*
+      LED push
+     */
+    void led_timer_tick(uint32_t last_run_us);
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/RCOutput.h:110]: uint64_t

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL_ChibiOS/Scheduler.cpp [hunk 9]
```diff
@@ -662,6 +662,7 @@ uint8_t Scheduler::calculate_thread_priority(priority_base base, int8_t priority
         { PRIORITY_CAN, APM_CAN_PRIORITY},
         { PRIORITY_TIMER, APM_TIMER_PRIORITY},
         { PRIORITY_RCOUT, APM_RCOUT_PRIORITY},
+        { PRIORITY_LED, APM_LED_PRIORITY},
```
Comment (tridge) [inline libraries/AP_HAL_ChibiOS/Scheduler.cpp:665]: might be needed in other HALs for completeness

**Review by tridge — CHANGES_REQUESTED** [review]

File: libraries/AP_HAL_ChibiOS/Scheduler.cpp (hunk 9 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/Scheduler.cpp:665]: Actually RCOUT is also missing so we clearly do not do this currently

File: libraries/AP_HAL_ChibiOS/RCOutput.h (hunk 8 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/RCOutput.h]: Fixed

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp (hunk 7 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/RCOutput.cpp]: Fixed

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp (hunk 6 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/RCOutput.cpp]: Fixed

File: libraries/AP_HAL_ChibiOS/RCOutput.cpp (hunk 5 above)
Comment (andyp1per) [reply libraries/AP_HAL_ChibiOS/RCOutput.cpp]: Fixed

**Pull request merged**